### DISCUSSION TITLE
How to store values that keep changing after every trial in data of a trial?

### DISCUSSION THREAD

**Discussion by prashantig25** [jspsych/jsPsych#1520] [Q&A]
Dear All,
I am trying to store data for each trial. In each trial, 2 gabor patches are displayed which have varying contrast levels. The contrast level changes after every trial through the on_start function. I want to store the contrast level for each trial in data. I am trying to do that by defining 'con' inside data and trying to store the value in it. But, when I run the experiment in Cognition.run, the column with 'con' is empty in the .csv file. Is there anyway I can fix it? Or is there any other way to store the contrast levels of each trial so that the data is stored for each trial? Here is part of the code:  

```
var s0_r1_high60 = {
        type: 'canvas-keyboard-response',
        on_start: function(s0_r1_high60){
            s0_r1_high60.contrast = Math.random(); s0_r1_high60.contrast1 = jsPsych.currentTrial().contrast  + randbet(0.05,0.35);
        },
        stimulus: gabor,
        data: {
            condition: 'HH', state: '0', reward: '1', prob: '60_40', con: jsPsych.currentTrial().contrast, 
        },
        choices: keys, trial_duration: duration,response_ends_trial: false,
        on_finish: function (data) {
            if (data.key_press === rand60r1.pop()) {
                data.correct = true;
            } else {
                data.correct = false;
            }
        },
    };
```
Any help is appreciated. Thank you!

**Comment by JAQuent** [comment]
I am not sure if I understand the problem but do you have to change anything as function of previous trials or can you pretty much set everything up before?  If the latter is the case, you can just create it in an array before:
```

var stimuli = new Array();
for (i = 0; i < nTrial; i++){
  stimuli[i] = {data: {condition: 'HH', 
                       state: '0', 
                       reward: '1', 
                       prob: '60_40', 
                       con: 'your contrast values here'}};
};

// Inside your s0_r1_high60 variable you set
data: jsPsych.timelineVariable('data'),


// Pushing it to the timeline
var experiment = {
  /// other arguments
  timeline_variables: stimuli,
  // other arguments
};
timeline.push(experiment);

```

**Reply by prashantig25** [reply]
Hi @JAQuent 
Thank you but I can't pre-define. The contrast level is supposed to be different for every trial and it needs to change on every trial. I do that using the on_start function. But I want to store that value in the data for every trial. I hope that is clear. 
Thanks.

**Comment by JAQuent** [comment]
Just to be clear. You can have the contrast level different every trial with that but what would not work with my suggestion is changing it as a function of previous trails (e.g. performance). Hope you'll find your solution. 

On 17 February 2021 20:45:45 GMT, prashantig25 <notifications@github.com> wrote:
>Hi @JAQuent 
>Thank you but I can't pre-define. The contrast level is supposed to be
>different for every trial and it needs to change on every trial. I do
>that using the on_start function. But I want to store that value in the
>data for every trial. I hope that is clear. 
>Thanks. 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/jspsych/jsPsych/discussions/1520#discussioncomment-379111

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by becky-gilbert** [comment]
Hi @prashantig25, sorry for the delayed response. I'm not sure if this is still helpful, but I thought I'd reply anyway.
I think the reason your method isn't working is because the `jsPsych.currentTrial` function that you have in the `data` parameter will run when the page first loads (before the experiment has started). So you could try making `data` a [dynamic parameter](https://www.jspsych.org/overview/dynamic-parameters/), i.e. a function that returns the data object.
```js
data: function() {
  return {condition: 'HH', state: '0', reward: '1', prob: '60_40', con: jsPsych.currentTrial().contrast};
}
```
If that doesn't work, another way to do this is to use a global variable to store the current contrast value during the experiment. You would change the value of this variable in the `on_start` function, and then access its value in the `data` parameter. Again, this means that your `data` parameter should be a function.
```js
var curr_contrast = null;

var s0_r1_high60 = {
  type: 'canvas-keyboard-response',
  on_start: function(s0_r1_high60){
    s0_r1_high60.contrast = Math.random(); s0_r1_high60.contrast1 = jsPsych.currentTrial().contrast  + randbet(0.05,0.35);
    curr_contrast = s0_r1_high60.contrast;
  },
  stimulus: gabor,
  data: function() {
    return {condition: 'HH', state: '0', reward: '1', prob: '60_40', con: curr_contrast};
  },
  // etc.
};
```
Finally, are you using a modified version of the `canvas-keyboard-response` plugin that has a `contrast` parameter? If so, then you could just modify the plugin so that it saves the `contrast` parameter value to the trial data by default. Or if you're using jsPsych v.6.3.0, then you can use the [`save_trial_parameters`](https://www.jspsych.org/overview/plugins/#the-save_trial_parameters-parameter) parameter, which lets you save (or not save) any of the trial parameters:
```js
var s0_r1_high60 = {
  type: 'canvas-keyboard-response',
  on_start: function(s0_r1_high60){
    s0_r1_high60.contrast = Math.random(); s0_r1_high60.contrast1 = jsPsych.currentTrial().contrast  + randbet(0.05,0.35);
    curr_contrast = s0_r1_high60.contrast;
  },
  stimulus: gabor,
  data: function() {
    return {condition: 'HH', state: '0', reward: '1', prob: '60_40'};
  },
  save_trial_parameters: {
    contrast: true  
  },
  // etc.
};
```
